File: src/Q57_2.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> twoSum(vector<int> &nums, int target) {
    vector<int> ret;
    for (int l = 0, r = nums.size() - 1; l < r;) {
      if (nums[l] + nums[r] == target) {
        ret.push_back(nums[l]);
        ret.push_back(nums[r]);
        break;
      } else if (nums[l] + nums[r] < target) {
        l++;
      } else {
        r--;
      }
    }
    return ret;
  }
};

File: src/Q57_2.rs
struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        if nums.len() <= 1 {
            return Vec::new();
        }
        use std::cmp::Ordering;
        let (mut l, mut r) = (nums.iter(), nums.iter().rev());
        let (mut lv, mut rv) = (l.next().unwrap(), r.next().unwrap());
        loop {
            match (lv + rv).cmp(&target) {
                Ordering::Equal => return vec![*lv, *rv],
                Ordering::Less => match l.next() {
                    Some(val) => lv = val,
                    _ => break,
                },
                Ordering::Greater => match r.next() {
                    Some(val) => rv = val,
                    _ => break,
                },
            }
        }
        Vec::new()
    }
}
